### DISCUSSION TITLE
Right aligned status icons merged to main!

### DISCUSSION THREAD

**Discussion by cseickel** [nvim-neo-tree/neo-tree.nvim#228] [Announcements]
This is something I've wanted from the beginning and I finally made it happen. With the new container support feature that I just merged to main, we now have right aligned diagnostic and git status icons. This also means that if there is not enough space, those icons will overlay the file name.
![image](https://user-images.githubusercontent.com/5160605/160962184-48263f9a-3f6d-490b-8635-7e81a5e31b1b.png)

Even better, when a file name is truncated the last three characters will now fade out to indicate that. This is my favorite part.
![image](https://user-images.githubusercontent.com/5160605/160962312-4719c912-1b72-4032-99e5-59ccfb6e5c1d.png)

https://user-images.githubusercontent.com/5160605/160962381-71cf9954-fc66-4f26-8564-c5b7a28b0b70.mp4

If you have configured custom renderers, you will need to copy the new container component config from the default config to take advantage of this.

I made one change to the default values, which is to remove the default staged/upstaged symbols from the git status component. I think it looks better that way because it keeps that component to a one character symbol all the time.

I'm going to let this sit in main for a few days to see if there are any issues. If you try it out, please provide feedback!

**Comment by gegoune** [comment]
<img width="245" alt="Screenshot 2022-03-31 at 10 43 39" src="https://user-images.githubusercontent.com/69750637/161014819-2c7c6117-4822-494b-b578-1cffa8a5c474.png">
Is that circle suppose to change somehow indicating staged/unstaged changes made to buffer?

**Reply by cseickel** [reply]
That circle indicates that the file has been modified. In the last commit, I set the default icons for stage and unstaged to an empty string, which effectively disables them.

![image](https://user-images.githubusercontent.com/5160605/161044236-1de17cb8-fbc6-4333-b7eb-03badb891ee9.png)

**This is open for debate, and I may change the defaults back before releasing these changes to the v2.x branch**.  My reasoning for this is that I just do not need this information on the tree. I always add my changes and commit in either the same command, or in two subsequent commands. There is never a time where I need to keep track of what was added and what was not.

Another variation I have used to make it one symbol is to go the opposite way and disable all of the "Change Type" icons, because if you use `git_status_colors = true` on the name component, which is the default, then that information is already contained in that color:

![image](https://user-images.githubusercontent.com/5160605/161045313-0ae0da72-ef71-480e-8beb-9a00431605cc.png)

What do you think? Are you in favor of keeping the current version of the one icon mix, the old default of sometimes two icons, or the one icon version that is all status and no change type indicators?

**Reply by gegoune** [reply]
I think that `staged/unstaged` could be bundled with `modified`, in a sense that in VCS' modified means unstaged unless it get staged so I probably would like to only see if file has staged changes or not (which would imply `modified`). That makes `modified` redundant, doesn't it? Or am I missing/misunderstanding anything here?

Edit: ah, of course modified means 'unsaved changes'. I think locally I will restore staged/unstaged icons.

**Reply by cseickel** [reply]
This should have been a reply instead of a new "answer", moving it here now:

It's mostly redundant. In the context of the filesystem source, a file may be either added (+) or modified (*), so that's the extra information that is lost by only using the status symbols.

In the context of the git_status source, there is also the additional deleted and renamed changed types that could show.

To be clear, it does not mean unsaved changes in the context of whether the changes have been saved to disk. These symbols are only for the git status.

**Reply by cseickel** [reply]
It sounds to me like you would be in favor of my second screenshot using only the status symbols, would you agree with that?

![image](https://user-images.githubusercontent.com/5160605/161099452-2388d07a-dc09-4f81-8662-e30ff06630c8.png)

I guess one more variation, which is not currently supported, is to show modified(*)/added(+) only if it is unstaged, and only the staged symbol (checkbox) if it is staged.

**Reply by gegoune** [reply]
Another option would be to maybe just display icons for each case? What if file is partially staged, modified and unsaved? :) I think there is many combinations and as much as I would like to help I am not sure if I hare enough insight to influence you in any way (worried my reasoning might be flowed). With that said, I like information and in theory wouldn't mind having all states covered. Perhaps type of icon would be locked to its column? There could be gaps between icons but scanning to see what's up should be pretty easy.

**Reply by cseickel** [reply]
> With that said, I like information and in theory wouldn't mind having all states covered. Perhaps type of icon would be locked to its column? There could be gaps between icons but scanning to see what's up should be pretty easy.

Now that's a good idea, and totally doable as a personal configuration change within what is deployed now. Maybe I will add a wiki page with screenshots and config example for all of these scenarios. I should also add a component to show if a file has unsaved changes.

I still need to figure out a sane default... maybe it's best to just revert the defaults to what it was in the last release.

**Comment by nyngwang** [comment]
@cseickel:

Where should I put the config to achieve the effect of the title? Sorry I didn't follow this thread the moment you were discussing it...

---

**update**: I just made it works by putting configs in `filesystem.renderers.file/directory`.

The question follows this is: It seems that

1. The usage of `'container'` is not documented in the README.md
2. The `renderers` can be moved outside of filesystem, but I don't know the reasoning. The one you used here seems to be the latest syntax, right?

**Reply by cseickel** [reply]
That is the default config, so if you are having any trouble just remove your overrides or copy from defaults.lua and make whatever changes you need from that base.

> 1. The usage of `'container'` is not documented in the README.md

The README is not the documentation, it is just the highlights for new users that have not installed the plugin yet. You should use the vim help file as your reference:
```
:h neo-tree-container
```



> 2\. The `renderers` can be moved outside of filesystem, but I don't know the reasoning. The one you used here seems to be the latest syntax, right?

There are two other sources and more coming eventually. Putting `renderers` at teh global level allows you to define it once for all sources instead of having to repeat it for each one. The same applies to the `window` section.

This does look like the final syntax.
